Deserialization of ics into Event Structs

johnhamelink opened the pull request:
This PR is designed to allow for this module to be able to parse ics strings into Event Structs to be read or manipulated in a standard format. This is useful if you are importing ics strings into a database, or if you want to edit an ics.

johnhamelink committed:
Add ability to deserialize events into Event structs

johnhamelink committed:
Remove unnecessary comments

johnhamelink committed:
Clean up dogma errors

johnhamelink committed:
Upgrade to elixir 1.2

johnhamelink committed:
Upgrade to 1.3.2

lpil commented:
Seems like a good feature.

lpil commented on lib/icalendar/deserialize.ex:
```diff
@@ -0,0 +1,71 @@
+      |> String.slice(4..5)
+      |> String.to_integer
+
+    {{year, month, day}, {hour, minute, second}}
```
Can timezone information be preserved?

johnhamelink replied:
Erlang timestamps don't have Timezones as far as I'm aware?

lpil replied:
If the format we're deserializing has timezone information it would be nice to preserve it. Can we return a data structure that does? Perhaps `DateTIme`?

johnhamelink replied:
Sorry I misread :smile: 

Yes, the timestamps could be in the following format: `DTSTART;TZID=US-Eastern:19980119T020000`. I chose not to implement this as we're not currently /producing/ dates like these. However that's where I want us to get to with #7

lpil commented on lib/icalendar/deserialize.ex:
```diff
@@ -0,0 +1,71 @@
+  def build_event({"SUMMARY", summary}, acc), do: %{acc | summary: summary}
+  def build_event(_, acc), do: acc
+
+  def to_date(date_string) do
```
Could we extract this function into a date handling module please? Might be nice to have a few tests to sanity check too.

johnhamelink committed:
Moved deserialization process into a utils library

johnhamelink committed:
Clean up to_date/1 add more tests

johnhamelink commented:
I've moved the deserialization process into a utils module. I wasn't happy with the rather verbose way I'd implemented it first, so I used named regex parameters to make it a bit cleaner.

I also added some doctests and moved some of the parsing-specific tests into the test suite for the deserialization module.

lpil commented on lib/icalendar/util/deserialize.ex:
```diff
@@ -0,0 +1,74 @@
+        time = {
+          String.to_integer(hour), String.to_integer(minute),
+          String.to_integer(second)}
+        {date, time}
```
This should be `{:ok, {date, time}}` as otherwise it's not obvious that an error tuple may be returned.

``` elixir
{date, time} = to_date(string) # This would match both a success and fail result.
```

lpil commented:
Nice. I've added one little comment :)

johnhamelink committed:
Use {:ok, erl_date} / {:error, msg} syntax for to_date/1

johnhamelink commented:
@lpil should be sorted now :+1:

lpil commented:
Ace, thanks!

lpil commented:
@johnhamelink Ah! This broke documentation generation :(

Do you think you could take a look at this please? Run the `mix docs` task to see what I mean. Also if you could add `mix docs` as a test to the travis.yml file so this can be caught that would be ace. Thanks

johnhamelink commented:
Will do, no problem :)

lpil commented:
Thank you.
